File: weave-cookbook/typescript/examples/03-llm-judge.ts
// An LLM judge for the one property code can't measure: does the pick match the vibe?

import * as weave from 'weave';

import { describeGame, filterGames, loadCatalog, type Game } from './shared/catalog.js';
import { inferenceClient, loadEnv, modelId, weaveProject } from './shared/config.js';
import { isDirectExecution } from './shared/runtime.js';

type CalibrationRow = { vibe: string; game: string; human_says: boolean };
type Scenario = { players: number; minutes: number; vibe: string };
type Verdict = { vibe_match: boolean; reason: string };

const RUBRIC =
  'You judge board game recommendations. Decide only whether the game matches the ' +
  'requested vibe. Judge mood and play style; ignore player count and duration. ' +
  'Return vibe_match and a one-sentence reason.';

const VERDICT_FORMAT = {
  type: 'json_schema' as const,
  json_schema: {
    name: 'VibeVerdict',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        vibe_match: { type: 'boolean' },
        reason: { type: 'string' },
      },
      required: ['vibe_match', 'reason'],
      additionalProperties: false,
    },
  },
};

// Human-labeled rows. The judge has to agree with these before its scores mean anything.
export const CALIBRATION: CalibrationRow[] = [
  { vibe: 'cooperative', game: 'Moonbase Delta', human_says: true },
  { vibe: 'cooperative', game: 'Caravan Kings', human_says: false },
  { vibe: 'quick party fun', game: 'Orchard Sprint', human_says: true },
  { vibe: 'quick party fun', game: 'Ironroot', human_says: false },
  { vibe: 'brainy deduction', game: "The Alchemist's Cellar", human_says: true },
  { vibe: 'brainy deduction', game: 'Sky Ferry', human_says: false },
];

const SCENARIOS: Scenario[] = [
  { players: 4, minutes: 60, vibe: 'cooperative' },
  { players: 2, minutes: 45, vibe: 'brainy deduction' },
  { players: 5, minutes: 30, vibe: 'quick party fun' },
];

const SYSTEM_PROMPT =
  'You are the Game Night Agent. Pick the best game for the group from the candidate list. ' +
  'Reply with the game name only, exactly as written in the list.';

function describe(gameName: string): string {
  const game = loadCatalog().find((entry: Game) => entry.name === gameName);
  return game ? describeGame(game) : gameName;
}

// Narrow LLM judge: one question, structured output.
const judgeVibe = weave.op(
  async function judge_vibe(vibe: string, game: string): Promise<Verdict> {
    const response = await weave.wrapOpenAI(inferenceClient()).chat.completions.create({
      model: modelId(),
      messages: [
        { role: 'system', content: RUBRIC },
        { role: 'user', content: `Requested vibe: ${vibe}\nRecommended game: ${describe(game)}` },
      ],
      response_format: VERDICT_FORMAT,
    });
    const content = response.choices[0].message.content;
    if (!content) {
      throw new Error('Judge returned no structured verdict.');
    }
    const verdict = JSON.parse(content) as Partial<Verdict>;
    if (typeof verdict.vibe_match !== 'boolean' || typeof verdict.reason !== 'string') {
      throw new Error('Judge verdict did not match the expected schema.');
    }
    return verdict as Verdict;
  },
);

const judgeModel = weave.op(
  async ({ datasetRow }: { datasetRow: CalibrationRow }) =>
    judgeVibe(datasetRow.vibe, datasetRow.game),
  { name: 'judge_vibe_model' },
);

const agreesWithHuman = weave.op(
  ({ modelOutput, datasetRow }: { modelOutput: Verdict; datasetRow: CalibrationRow }) => {
    return { agrees_with_human: modelOutput.vibe_match === datasetRow.human_says };
  },
  { name: 'agrees_with_human' },
);

// Scorer that delegates to the judge. Not ground truth -- a calibrated opinion.
const vibeMatch = weave.op(
  async ({ modelOutput, datasetRow }: { modelOutput: string; datasetRow: Scenario }) => {
    const verdict = await judgeVibe(datasetRow.vibe, modelOutput);
    return { vibe_match: verdict.vibe_match, judge_reason: verdict.reason };
  },
  { name: 'vibe_match' },
);

const recommend = weave.op(
  async ({ datasetRow }: { datasetRow: Scenario }) => {
    const candidates = filterGames(loadCatalog(), {
      players: datasetRow.players,
      maxMinutes: datasetRow.minutes,
    });
    if (candidates.length === 0) {
      return 'none';
    }
    const prompt =
      `Players: ${datasetRow.players}\n` +
      `Time available: ${datasetRow.minutes} minutes\n` +
      `Vibe: ${datasetRow.vibe}\n\n` +
      `Candidates:\n${candidates.map(describeGame).join('\n')}`;
    const response = await weave.wrapOpenAI(inferenceClient()).chat.completions.create({
      model: modelId(),
      messages: [
        { role: 'system', content: SYSTEM_PROMPT },
        { role: 'user', content: prompt },
      ],
    });
    return (response.choices[0].message.content ?? '').trim().replace(/^"|"$/g, '');
  },
  { name: 'recommend' },
);

export async function main(): Promise<void> {
  loadEnv();
  await weave.init(weaveProject());

  const calibration = new weave.Evaluation({
    name: 'vibe-judge-calibration',
    dataset: new weave.Dataset({ name: 'vibe-judge-calibration-rows', rows: CALIBRATION }),
    scorers: [agreesWithHuman],
  });
  const summary = await calibration.evaluate({ model: judgeModel });
  const agreement = summary.agrees_with_human.agrees_with_human.true_fraction;
  console.log(
    `Judge agrees with the human labels on ${Math.round(agreement * 100)}% of calibration rows.`,
  );

  const agentEval = new weave.Evaluation({
    name: 'vibe-judge-eval',
    dataset: new weave.Dataset({ name: 'vibe-scenarios', rows: SCENARIOS }),
    scorers: [vibeMatch],
  });
  const agentSummary = await agentEval.evaluate({ model: recommend });
  console.log(agentSummary.vibe_match.vibe_match);
}

if (isDirectExecution(import.meta.url)) {
  await main();
}

File: weave-cookbook/python/examples/03_llm_judge.py
"""An LLM judge for the one property code can't measure: does the pick match the vibe?"""

import asyncio
import json

import weave

from shared.catalog import describe_game, filter_games, load_catalog
from shared.config import inference_client, model_id, require_api_key, weave_project

RUBRIC = (
    "You judge board game recommendations. Decide only whether the game matches the "
    "requested vibe. Judge mood and play style; ignore player count and duration. "
    "Return vibe_match and a one-sentence reason."
)

VERDICT_FORMAT = {
    "type": "json_schema",
    "json_schema": {
        "name": "VibeVerdict",
        "strict": True,
        "schema": {
            "type": "object",
            "properties": {
                "vibe_match": {"type": "boolean"},
                "reason": {"type": "string"},
            },
            "required": ["vibe_match", "reason"],
            "additionalProperties": False,
        },
    },
}

# Human-labeled rows. The judge has to agree with these before its scores mean anything.
CALIBRATION = [
    {"vibe": "cooperative", "game": "Moonbase Delta", "human_says": True},
    {"vibe": "cooperative", "game": "Caravan Kings", "human_says": False},
    {"vibe": "quick party fun", "game": "Orchard Sprint", "human_says": True},
    {"vibe": "quick party fun", "game": "Ironroot", "human_says": False},
    {"vibe": "brainy deduction", "game": "The Alchemist's Cellar", "human_says": True},
    {"vibe": "brainy deduction", "game": "Sky Ferry", "human_says": False},
]

SCENARIOS = [
    {"players": 4, "minutes": 60, "vibe": "cooperative"},
    {"players": 2, "minutes": 45, "vibe": "brainy deduction"},
    {"players": 5, "minutes": 30, "vibe": "quick party fun"},
]

SYSTEM_PROMPT = (
    "You are the Game Night Agent. Pick the best game for the group from the candidate list. "
    "Reply with the game name only, exactly as written in the list."
)


def describe(game_name: str) -> str:
    for game in load_catalog():
        if game["name"] == game_name:
            return describe_game(game)
    return game_name


@weave.op()
def judge_vibe(vibe: str, game: str) -> dict:
    """Narrow LLM judge: one question, structured output."""
    response = inference_client().chat.completions.create(
        model=model_id(),
        messages=[
            {"role": "system", "content": RUBRIC},
            {
                "role": "user",
                "content": f"Requested vibe: {vibe}\nRecommended game: {describe(game)}",
            },
        ],
        response_format=VERDICT_FORMAT,
    )
    content = response.choices[0].message.content
    if not content:
        raise ValueError("Judge returned no structured verdict.")
    verdict = json.loads(content)
    if not isinstance(verdict.get("vibe_match"), bool) or not isinstance(
        verdict.get("reason"), str
    ):
        raise ValueError("Judge verdict did not match the expected schema.")
    return verdict


@weave.op()
def agrees_with_human(human_says: bool, output: dict) -> dict:
    return {"agrees_with_human": output["vibe_match"] == human_says}


@weave.op()
def vibe_match(vibe: str, output: str) -> dict:
    """Scorer that delegates to the judge. Not ground truth -- a calibrated opinion."""
    verdict = judge_vibe(vibe, output)
    return {"vibe_match": verdict["vibe_match"], "judge_reason": verdict["reason"]}


@weave.op()
def recommend(players: int, minutes: int, vibe: str) -> str:
    candidates = filter_games(load_catalog(), players=players, max_minutes=minutes)
    if not candidates:
        return "none"
    prompt = (
        f"Players: {players}\n"
        f"Time available: {minutes} minutes\n"
        f"Vibe: {vibe}\n\n"
        "Candidates:\n" + "\n".join(describe_game(game) for game in candidates)
    )
    response = inference_client().chat.completions.create(
        model=model_id(),
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": prompt},
        ],
    )
    return (response.choices[0].message.content or "").strip().strip('"')


def main() -> None:
    require_api_key()
    weave.init(weave_project())

    calibration = weave.Evaluation(
        dataset=CALIBRATION,
        scorers=[agrees_with_human],
        evaluation_name="vibe-judge-calibration",
    )
    summary = asyncio.run(calibration.evaluate(judge_vibe))
    agreement = summary["agrees_with_human"]["agrees_with_human"]["true_fraction"]
    print(f"Judge agrees with the human labels on {agreement:.0%} of calibration rows.")

    agent_eval = weave.Evaluation(
        dataset=SCENARIOS,
        scorers=[vibe_match],
        evaluation_name="vibe-judge-eval",
    )
    agent_summary = asyncio.run(agent_eval.evaluate(recommend))
    print(agent_summary["vibe_match"]["vibe_match"])


if __name__ == "__main__":
    main()
